copycat/api/src/main/java/org/apache/kafka/copycat/data/SchemaProjector.java

```diff
@@ -0,0 +1,209 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * <p>
+ *     SchemaProjector is utility to project a value between compatible schemas and throw exceptions
+ *     when non compatible schemas are provided.
+ * </p>
+ */
+
+public class SchemaProjector {
+
+    /**
+     * @param writer the schema that is used to write the record
+     * @param reader the schema that is used to read the record
+     * @param record the value to project from writer schema to reader schema
+     * @return the projected value with reader schema
+     * @throws SchemaProjectorException
+     */
+    public static Object project(Schema writer, Schema reader, Object record) throws SchemaProjectorException {
+        if (writer.isOptional() && !reader.isOptional()) {
+            if (reader.defaultValue() != null) {
+                if (record != null) {
+                    return projectRequiredSchema(writer, reader, record);
+                } else {
+                    return reader.defaultValue();
+                }
+            } else {
+                throw new SchemaProjectorException("Writer schema is optional, however, reader schema does not provide a default value.");
+            }
+        } else {
+            if (record != null) {
+                return projectRequiredSchema(writer, reader, record);
+            }
+            return null;
+        }
+    }
+
+    private static Object projectRequiredSchema(Schema writer, Schema reader, Object record) throws SchemaProjectorException {
+        Object result = null;
+        // handle logical types
+        if (reader.name() != null) {
+            return projectLogical(writer, reader, record);
+        }
+        switch (reader.type()) {
+            case INT8:
+            case INT16:
+            case INT32:
+            case INT64:
+            case FLOAT32:
+            case FLOAT64:
+            case BOOLEAN:
+            case BYTES:
+            case STRING:
+                result = projectPrimitive(writer, reader, record);
+                break;
+            case STRUCT:
+                result = projectStruct(writer, reader, (Struct) record);
+                break;
+            case ARRAY:
+                result = projectArray(writer, reader, record);
+                break;
+            case MAP:
+                result = projectMap(writer, reader, record);
+        }
+        return result;
+    }
+
+    private static Object projectStruct(Schema writer, Schema reader, Struct writerStruct) throws SchemaProjectorException {
+        Struct readerStruct = new Struct(reader);
+        for (Field readerField : reader.fields()) {
+            String fieldName = readerField.name();
+            Field writerField = writer.field(fieldName);
+            if (writerField != null) {
+                Object writerFieldValue = writerStruct.get(fieldName);
+                Object readerFieldValue = project(writerField.schema(), readerField.schema(), writerFieldValue);
```

Comment:
It's not critical, but you might want to catch any exceptions from this `project` call and wrap them in another `SchemaProjectorException` so you can include info about which field failed projection.